pcsx2/SaveState.cpp

```diff
@@ -1144,13 +1144,22 @@ bool SaveState_UnzipFromDisk(const std::string& filename, Error* error)
 	}
 
 	PreLoadPrep();
+	const bool extraMemory = CHECK_EXTRAMEM;
 
 	if (!LoadInternalStructuresState(zf.get(), internal_index))
 	{
 		Error::SetString(error, "Save state corruption in internal structures.");
 		return false;
 	}
 
+	// A save state is not allowed to change the extra memory mode.
+	if (extraMemory != CHECK_EXTRAMEM)
+	{
+		memSetExtraMemMode(extraMemory);
```

Comment:
I've pushed a first attempt at doing this. I ended up using `zip_fseek` to get the size of the entry because relying on `bytesRead` wouldn't handle loading a 128MB state in 32MB mode, since there's always going to be 32MB available (therefore it will succeed and not fail to load the state).

It feels like `MemorySavestateEntry::FreezeIn` should actually be doing this validation instead of it being a special case in `SavestateEntry_EmotionMemory`, especially since it actually already warns for an incomplete file but continues loading the state anyway. What are your thoughts?

Aside from that, the main downside currently is that there is no longer a message telling you that you need to enable or disable expanded memory. If you try to load a 128MB state in 32MB mode or vice versa, you will just get a message about corruption in the save state since there is no way to specify the error message from `FreezeIn` currently.

Reply:
I'll see about adding error propagation.